test/metabase/db/schema_migrations_test/impl.clj

```diff
@@ -200,33 +206,41 @@
   (log/debug (u/format-color 'yellow "Testing migrations for driver %s..." driver))
   (with-temp-empty-app-db [conn driver]
     ;; sanity check: make sure the DB is actually empty
-    (let [metadata (.getMetaData conn)
-          schema (when (= :h2 driver) "PUBLIC")]
+    (let [metadata  (.getMetaData conn)
+          schema    (when (= :h2 driver) "PUBLIC")]
       (with-open [rs (.getTables metadata nil schema "%" (into-array String ["TABLE"]))]
         (let [tables (jdbc/result-set-seq rs)]
           (assert (zero? (count tables))
                   (str "'Empty' application DB is not actually empty. Found tables:\n"
                        (u/pprint-to-str tables))))))
     (log/debugf "Finding and running migrations before %s..." start-id)
     (run-migrations-in-range! conn ["v00.00-000" start-id] {:inclusive-end? false})
-    (letfn [(migrate
-              ([]
-               (migrate :up nil))
-              ([direction]
-               (migrate direction nil))
+    (let [restart-id (atom nil)]
+      (letfn [(migrate
+                ([]
+                 (migrate :up nil))
+                ([direction]
+                 (migrate direction nil))
 
-              ([direction version]
-               (case direction
-                 :up
-                 (do
-                  (log/debugf "Finding and running migrations between %s and %s (inclusive)" start-id (or end-id "end"))
-                  (run-migrations-in-range! conn [start-id end-id]))
+                ([direction version]
+                 (case direction
+                   :up
+                   (do
+                     (log/debugf "Finding and running migrations between %s and %s (inclusive)" start-id (or end-id "end"))
+                     ;; If we have rolled back earlier migrations, it's no longer safe to resume from start-id.
+                     (if-let [start-after @restart-id]
+                       (run-migrations-in-range! conn [start-after end-id] {:inclusive-start? false})
+                       (run-migrations-in-range! conn [start-id end-id])))
```

Comment:
Small thing: this logging is now out of date. It should be conditional on `start-after`:
```
                     (if-let [start-after @restart-id]
                       (do
                         (log/debugf "Finding and running migrations after %s until %s (inclusive)" start-after (or end-id "end"))
                         (run-migrations-in-range! conn [start-after end-id] {:inclusive-start? false}))
                       (do
                         (log/debugf "Finding and running migrations between %s and %s (inclusive)" start-id (or end-id "end"))
                         (run-migrations-in-range! conn [start-id end-id]))))
```

Reply:
Great catch, will send a new PR to fix